core/service/src/lib.rs

```diff
@@ -746,19 +761,27 @@ fn build_network_future<
 		}
 
 		// Interval report for the external API.
-		while let Ok(Async::Ready(_)) = status_interval.poll() {
-			let status = NetworkStatus {
-				sync_state: network.sync_state(),
-				best_seen_block: network.best_seen_block(),
-				num_sync_peers: network.num_sync_peers(),
-				num_connected_peers: network.num_connected_peers(),
-				num_active_peers: network.num_active_peers(),
-				average_download_per_sec: network.average_download_per_sec(),
-				average_upload_per_sec: network.average_upload_per_sec(),
-			};
-			let state = network.network_state();
-
-			status_sinks.lock().retain(|sink| sink.unbounded_send((status.clone(), state.clone())).is_ok());
+		{
+			let mut status_sinks = status_sinks.lock();
+			for n in (0..status_sinks.len()).rev() {
+				while let Ok(Async::Ready(_)) = status_sinks[n].0.poll() {
+					let status = NetworkStatus {
```

Comment:
`state` and `status` should be build outside of the for loop, only once.

Reply:
I went for creating the state inside the loop, because that loop is called every single time something happens on the network (i.e. very very often). In practice, the inside of the loop will only be reached three times every 5 seconds.

Reply:
You still call clone below.